Core/src/RequestWrapper.php

```diff
@@ -282,38 +283,41 @@ private function applyHeaders(RequestInterface $request)
         ];
 
         if ($this->shouldSignRequest) {
-            $headers['Authorization'] = 'Bearer ' . $this->getToken();
-        }
+            $quotaProject = $this->quotaProject;
+            $token = null;
+
+            if ($this->accessToken) {
+                $token = $this->accessToken;
+            } else {
+                $credentialsFetcher = $this->getCredentialsFetcher();
+                $token = $this->fetchCredentials($credentialsFetcher);
+                $quotaProject = $credentialsFetcher->getQuotaProject() ?: $quotaProject;
+            }
 
-        return Psr7\modify_request($request, ['set_headers' => $headers]);
-    }
+            $headers['Authorization'] = 'Bearer ' . $token;
 
-    /**
-     * Gets the access token.
-     *
-     * @return string
-     */
-    private function getToken()
-    {
-        if ($this->accessToken) {
-            return $this->accessToken;
+            if ($this->quotaProject) {
+                $headers[GetQuotaProjectInterface::X_GOOG_USER_PROJECT_HEADER] = [$quotaProject];
+            }
         }
 
-        return $this->fetchCredentials()['access_token'];
+
+        return Psr7\modify_request($request, ['set_headers' => $headers]);
     }
 
     /**
      * Fetches credentials.
      *
+     * @param FetchAuthTokenInterface $credentialsFetcher
      * @return array
      */
-    private function fetchCredentials()
+    private function fetchCredentials(FetchAuthTokenInterface $credentialsFetcher)
```

Comment:
LGTM, but curious why? unrelated refactoring?

Reply:
It's related :) - we needed access to the credentials class in the upper scope in order to get quota project.